### ISSUE TITLE
data_structure.py

### ISSUE THREAD

**Issue by ly29** [nortikin/sverchok#1086]
Reorganising the oldest utility file or ...
*The slaughter!*

Seriously there is a certain amount of cruft accumulated into here and some internal interfaces. After #1076 is merged I will review this file.

Things to be kept for now, 
- functions actually dealing with the data structure of Sverchok.
- old socket interface

Things to be moved
- code used by one node will either be moved into one the node or into a utils on a per case basis
- socket data cache into core
- other

Things to be removed
- Unused things, of which there are more than it should be...

Contents of ```data_structure```
- ~~Cnodes~~, some sort of caching system, not used
- ~~bmesh mapping~~, a dictionary for bmesh objects, not used
- "cache magic" an interface for a dictionary, used by object in (only?), move
- "list match magic", widely used, especially ```match_long_repeat```, keep
- the decorators, almost magical nodes, keep?, moved to end for now...
- data level things with data correct, etc. long term this needs to be reworked, but as part of a bigger change, keep
- vectors and matrix support functions to and from Sv data, to be kept
- ~~"list magic"~~, should be moved into a util I think, used by list zip and list join
- ```enum_item``` should perhaps be kept.
- debug settings, keep for now...
- heat map, moved into ui?
- socket handling functions, to be kept for now, could be refactored into classes
- socket_id and node_id, should be moved into socket and node classes
- ~~Socket data cache~~, old interface, kept but gradually replaced, ```SvGet/SetSocketAnyType```, will be redirected to new interface in the meantime
- ~~socket data cache~~, implementation, will be moved out of file
- ~~svQsort~~, unsure, I don't see any reason to use this instead of standard ```sorted``` with ```key``` function

Comments? @zeffii @nortikin @Kosvor2

**Comment by ly29** [comment]
Also all functions that are kept will get a doc string...

**Comment by ly29** [comment]
```python
def svQsort(L):
    if L: return svQsort([x for x in L[1:] if x<L[0]]) + L[0:1] + svQsort([x for x in L[1:] if x>=L[0]])
    return []
```
Any reason to use this instead of ```sorted``` ?

```
>>> svQsort([3,2,1])
[1, 2, 3]

>>> sorted([3,2,1])
[1, 2, 3]
```

**Comment by zeffii** [comment]
can't comment at length at the moment but will later.

---
By far the most resonating comment there is that all function should have a docstring. _some_ form of accurate documentation.  Even the more obvious functions, just so something like oxydoc / sphynx could convert these py files into readable docs.. 

There are functions in that file that i've never tried to mentally-compile

enum_item  is something that I was thinking should probably go into a properties helper module, somewhere where we could even have property generator functions that return StringProperty() with update=updateNode automatically added.  like `UStringProperty(.....)

**Comment by zeffii** [comment]
- regarding the decorator for bmesh, am not sure if any node even uses it. It was a nice idea but got too little traction when it was proposed by @portnov .

While not directly in data_structure.py , the object_in node is the only one that uses one of its structures.. hence I still want to sink time into finding a solution. 
- objects_in is not working on my machine (due to C change introduced in mid November 2016) , but that node should not be writing / seeking any shared data from within the `buttons_draw` function, it get called too often on repeat and it's probably not needed. I think the sentiment is ' if you're doing anything other than drawing ui from inside `draw_buttons`, expect stuff to break. '

**Comment by ly29** [comment]
@zeffii 
The sane thing since it is a quite limited number of objects one would use here is to use a ```CollectionProperty``` with a ```PropertyGroup``` of string name

**Comment by zeffii** [comment]
like : https://github.com/nortikin/sverchok/blob/master/nodes/script/multi_exec.py#L40-L63

**Comment by ly29** [comment]
Something like that

**Comment by ly29** [comment]
So I propose, without much research to simply follow [PEP 257](https://www.python.org/dev/peps/pep-0257/) 
Or perhaps something more structured like [numpy docstring](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Simpler is better?

**Comment by ly29** [comment]
```python
def create_list(x, y):
    if type(y) in [list, tuple]:
        return reduce(create_list, y, x)
    else:
        return x.append(y) or x
```
Now this is just devious, consider:
```
>>> x=[2,4,5]
>>> x.append(5) or x
[2, 4, 5, 5]
```
Since the append will return ```None``` which evals to false the second part will always be executed making this effectively the same as:
```python
def create_list(x, y):
    if isinstance(y, (tuple, list)):
        return reduce(create_list, y, x)
    else:
        x.append(y)
        return x
```
However ```reduce``` is hard to follow for me.

**Comment by zeffii** [comment]
seen as that would be part of the algorithm, i prefer not to use the side-effect of the append returning None therefore the value after the `or` being returned for `create_list`. Recursion is tricky enough :)

-- admittedly it does look neat :) so I can go either way on it..

**Comment by ly29** [comment]
Neat, but is it worth the effort to save one line?
Once you understand it makes sense but it does make you blink once or twice, pythonic it ain't.

**Comment by ly29** [comment]
Right now calling a halt, now it needs testing and fixing things...

From 1209 lines to 779 right now.

**Comment by ly29** [comment]
So fixed errors from loading and shuffled some more.

**Comment by ly29** [comment]
Okay it needs some testing I think I am done with this for now.

**Comment by nortikin** [comment]
it is part of list level and list zip. it reduce level or increase level of list, as i understand. much old code

**Comment by zeffii** [comment]
@ly29    maybe 
```python
def replace_socket(socket, new_type, new_name=None, new_pos=None):
    '''
    Replace a socket and keep links
    '''
    if new_name is None:
        new_name = socket.name
    socket.name = new_name
    # quit early
    #if socket.bl_idname == new_type:
    #    return socket
    ng = socket.id_data
    ng.freeze()
    if socket.is_output:
        to_sockets = [l.to_socket for l in socket.links]
        outputs = socket.node.outputs
        if new_pos is None:
            for i,s in enumerate(outputs):
                if s == socket:
                    node_pos = i
                    break
        else:
            node_pos = new_pos

        outputs.remove(socket)
        new_socket = outputs.new(new_type, new_name)
        outputs.move(len(outputs)-1, node_pos)
        for to_socket in to_sockets:
            ng.links.new(new_socket, to_socket)
    else:
        if socket.is_linked:
            from_socket = socket.links[0].from_socket
        else:
            from_socket = None
        inputs = socket.node.inputs
        if new_pos is None:
            for i,s in enumerate(inputs):
                if s == socket:
                    node_pos = i
                    break
        else:
            node_pos = new_pos
        inputs.remove(socket)
        new_socket = inputs.new(new_type, new_name)
        inputs.move(len(inputs)-1, node_pos)
        if from_socket:
            ng.links.new(from_socket, new_socket)
    ng.unfreeze()
    return new_socket
```

can be part of SocketCommon

```python
class SvSocketCommon:
    @property
    def other(self):
        return get_other_socket(self)

    def set_default(self, value):
        if self.prop_name:
            setattr(self.node, self.prop_name, value)

    @property
    def index(self):
        node = self.node
        sockets = node.outputs if self.is_output else node.inputs
        for i, s in enumerate(sockets):
             if s == self:
                 return i

    def sv_set(self, data):
        SvSetSocket(self, data)

    def replace(self, new_type, new_name=None, new_pos=None):
        replace_socket(self, new_type, new_name, new_pos)
```

**Comment by zeffii** [comment]
that would make this more cohesive
```
      outputs.remove(outputs[0])
      outputs.new('StringsSocket', 'Noise S')
```
as
```python
     outputs[0].replace('StringsSocket', 'Noise S')
```

maybe a cut-down version of replace socket
```
    replace(self, new_type, new_name=None):
        self.bl_idname = new_type
        self.name = new_name or self.name
```

**Comment by zeffii** [comment]
hmm, why do seem to recall being able to write `self.bl_idname = new_type`

**Comment by ly29** [comment]
because it would be very nice if it was that simple.

It would be a logical and good way of exposing that data.

**Comment by zeffii** [comment]
```python
    def replace(self, new_type, new_name=None, new_pos=None):
        replace_socket(self, new_type, new_name, new_pos=new_pos or self.index)
```

**Comment by ly29** [comment]
Merged, I think I have it under control but any fixes should be small and quick. 
In that case sorry.
